Porting Clawback documentation to xrpl.org (#2111)
XRPLF/xrpl-dev-portal: content/concepts/tokens/clawback-of-issued-currency.md

---

> html: clawback-of-issued-currency.html

Reviewer: Doesn't match the h1 of the page. Now, this one is tricky because "Clawback" is also the exact name of the transaction so it can't be the same.

But for style reasons, we prefer to use "tokens" rather than "issued currency". So maybe:

```suggestion
html: clawing-back-tokens.html
```

And then change the h1 on line 9 to be "Clawing Back Tokens" to match. Bonus points, this has a nice parallel to "Freezing Tokens".

Might also be worth adding a note to the style guide to use "claw back" (two words) as a verb and "clawback" (one word) as a noun.

Reviewer 2: Thank you, Rome, for your thorough review.

parent: tokens.html

> blurb: Issuers can claw back their issued tokens for compliance purposes.

Reviewer: There's no need to say "issued tokens" in most places; maybe "fungible tokens" if you want to distinguish from NFTs, but "tokens" is enough in most contexts. Also, this makes the feature sound scarier than it is and misses one of the big restrictions on this feature.

```suggestion
blurb: Issuers can claw back their tokens for compliance purposes if they enable the Clawback feature before issuing tokens.
```

Reviewer 2: Thank you, Rome, for your thorough review.

labels:
  - Tokens
status: not_enabled
---
# Clawback

> <embed src="/snippets/_clawback-disclaimer.md" />

Reviewer: This needs to be replaced with Dactyl-compatible syntax _and_ you need to actually commit the snippet to this repo.

```suggestion
{% include '_snippets/_clawback-disclaimer.md' %}
```

Reviewer 2: Thank you, Rome, for your thorough review.

> For regulatory purposes, some issuers must have the ability to recover issued tokens after they are distributed to accounts. For example, if an issuer were to discover that tokens were sent to an account sanctioned for illegal activity, the issuer could recover, or *claw back*, the funds.

Reviewer: ```suggestion
For regulatory purposes, some issuers need the ability to recover tokens after they are distributed to accounts. For example, if an issuer were to discover that tokens were sent to an account sanctioned for illegal activity, the issuer could recover, or *claw back*, the funds.
```

Reviewer 2: Thank you, Rome, for your thorough review.

> Issuers can create tokens that they can claw back by adding the `lsfAllowTrustLineClawback` flag to the issuing account.

Reviewer: This sounds like setting the flag is how you create tokens. Also, I suggest giving the flag a prose name that isn't tied to a prefix, since it's conceptually the same setting whether you're discussing the `asf` or `lsf` values and names.

```suggestion
Issuers can gain the ability to claw back their tokens by enabling the **Allow Clawback** flag on their issuing account. This flag cannot be enabled if the issuer has already issued tokens.
```

Based on the code, the name has been shortened from "Allow Trust Line Clawback" to just "Allow Clawback" (in camel case with the relevant `asf`/`lsf` prefix regardless).

Reviewer 2: Thank you, Rome, for your thorough review.

Looking at TxFlags.h, the line item still seems to indicate asfAllowTrustLineClawback as the flag name. Where are you seeing code that has a different name?

constexpr std::uint32_t asfAllowTrustLineClawback          = 16;

Reviewer 2: Checking with the Tokenization team, it's actually the opposite - they _added_ Trust Line to the commands.

> **Note:** You can only claw back issued tokens created by your account. You cannot claw back XRP in this way.

Reviewer: ```suggestion
**Note:** You can only claw back fungible tokens issued by your account. You cannot claw back XRP or NFTs in this way.
```

> Clawback is disabled by default. To use clawback, you must set the `lsfAllowTrustLineClawback` flag through an `AccountSet` transaction. The `AccountSet` transaction only succeeds if the account has an empty owner directory, meaning that the account has no trust lines, offers, escrows, payment channels, or checks. After you set this flag, it cannot reverted. The account permanently gains the ability to claw back issued assets on trust lines.

Reviewer: ```suggestion
Clawback is disabled by default. To use clawback, you must send an [AccountSet transaction][] to enable the **Allow Clawback** setting. **An issuer with any existing tokens cannot enable Clawback.** You can only enable Allow Clawback if you have a completely empty owner directory, meaning you must do so before you set up any trust lines, offers, escrows, payment channels, checks, or even signer lists.
```

Note, the reference-style link I use for AccountSet transaction won't work unless you include the common links snippets at the end of the file.

Reviewer 2: Thank you, Rome, for your thorough review.

If you attempt to set `lsfAllowTrustLineClawback` while `lsfNoFreeze` is set, the transaction returns `tecNO_PERMISSION`, because clawback cannot be enabled on an account that has already disclaimed the ability to freeze trust lines. 

> Conversely, if you try to set `lsfNoFreeze` while `lsfAllowTrustLineClawback` is set, the transaction also returns `tecNO_PERMISSION`.

Reviewer: Rather than referring to the `lsf` flag, use the prose names for both **Allow Clawback** and **No Freeze**, and link [the relevant docs for No Freeze](https://xrpl.org/freezes.html#no-freeze).

Also, putting these sentences on different lines is pointless. They'll be treated as one hard-wrapped continuation of the same paragraph, so they should be on one line, unless the intent is for them to be separate paragraphs, in which case you need a blank line between them.

I also think you can more clearly summarize the rule up-front.

```suggestion
Allow Clawback is incompatible with the [No Freeze setting](freezes.html#no-freeze). You cannot enable Allow Clawback on an account that already has No Freeze enabled, and you cannot enable No Freeze on an account that already has Allow Clawback enabled.
```

## Example Clawback Transaction

```json
{
  "TransactionType": "Clawback",
  "Account": "rp6abvbTbjoce8ZDJkT6snvxTZSYMBCC9S",
  "Amount": {
      "currency": "FOO",
      "issuer": "rsA2LpzuawewSBQXkiju3YQTMzW13pAAdW",
      "value": "314.159"
    }
}
```

> In execution, this transaction would claw back at most 314.159 FOO issued by rp6abvbTbjoce8ZDJkT6snvxTZSYMBCC9S and held by rsA2LpzuawewSBQXkiju3YQTMzW13pAAdW. If rsA2LpzuawewSBQXkiju3YQTMzW13pAAdW does not have a trust line set up or that trust line's balance is 0, then the error `tecNO_LINE` is returned and a fee is consumed.

Reviewer: ```suggestion
If successful, this transaction would claw back at most 314.159 FOO issued by rp6abvbTbjoce8ZDJkT6snvxTZSYMBCC9S and held by rsA2LpzuawewSBQXkiju3YQTMzW13pAAdW.
```

Reviewer 2: Thank you, Rome, for your thorough review.
